src/action/filesystems.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the filesystems action.
+//!
+//! The filesystems action lists all mounted filesystems on the client,
+//! collecting device name, mount point, filesystem type and its options.
+//! Current implementation works only in Linux systems.
+
+use rrg_proto::{Filesystem, KeyValue, AttributedDict, DataBlob};
+use crate::session::{self, Session};
+
+/// A response type for the filesystems action.
+pub struct Response {
+    /// Information about filesystem.
+    mount_info: proc_mounts::MountInfo,
+}
+
+/// Handles requests for the filesystems action.
+/// Initially searches in `/proc/mounts`. If it's missed, falls back to
+/// `/etc/mtab`.
+pub fn handle<S: Session>(session: &mut S, _: ()) -> session::Result<()> {
+    use proc_mounts::MountIter;
+
+    let mount_iter = match MountIter::new() {
+        Ok(mount_iter) => mount_iter,
+        Err(_) => {
+            // /etc/mtab must exist for sure.
+            MountIter::new_from_file("/etc/mtab").expect("/etc/mtab not found")
```

Comment:
Ok, I think I need to add some custom `Error` structs, that handle all this kinds of errors, which can happen while my action is running on the client (missed `/etc/mtab`, `None` value in `Option<MountInfo>` during iteration in a loop and, lastly, invalid syntax of mount options).

Do I have to implement separate `Error` structs for handling all this errors or just one for all of them would be enough?

Reply:
For sure you will need to have one root `Error` type. If most cases are simple, then having them as enum variants with some payload should be enough. But if they need to carry more structured data, some of them might need to be refactored to separate types. You can take a look at `src/session/error.rs` for rough guidance of what level of granularity I am talking about here.

It can be a lot of boilerplate, but it should be simple and easy to write.

Reply:
I made `Error` enum with two variants: `MissedFile` and `MountInfoError`.
Can I see somehow in GRR server ui my error returned from `handle` function? When I intentionally specified unexistent files, I didn't find my error anywhere except general message "Could not enumerate file systems" in Log page.